src/thir.md

```diff
@@ -4,10 +4,11 @@
 
 The THIR ("Typed High-Level Intermediate Representation"), previously called HAIR for
 "High-Level Abstract IR", is another IR used by rustc that is generated after
-[type checking]. It is (as of <!-- date: 2021-03 --> March 2021) only used for
-[MIR construction] and [exhaustiveness checking], but
-[it may also soon be used for unsafety checking][thir-unsafeck] as a replacement
-for the current MIR unsafety checker.
+[type checking]. It is (as of <!-- date: 2021-08 --> August 2021) only used for
+[MIR construction] and [exhaustiveness checking]. There is also
+[an experimental unsafety checker][thir-unsafeck] that operates on the THIR as a replacement for
```

Comment:
I'm curious why there is an unsafety checker in THIR that aims to replace the one in MIR. Could we add something about that?

Reply:
See rust-lang/compiler-team#402. TL;DR is that unsafety checking is expected to match the syntax, but MIR doesn't match syntax closely which led to issues like
```rs
// deref of raw ptr should be unsafe even though the result is unused,
// but passes with MIR unsafeck
*ptr;
```

We can still handle cases like this in MIR (by using e.g. fake reads), but checking unsafety in THIR should be more natural and require less/no hacks.